src/parse_error.rs

```diff
@@ -138,3 +120,197 @@ impl std::fmt::Display for ParseErrorDisplay<'_> {
+    loop {
+        let node = cursor.node();
+        if node.is_error() {
+            errors.push(ParseError::Unexpected(node));
+            if first_only {
+                break;
+            }
+            did_visit_children = true;
+        } else if node.is_missing() {
+            errors.push(ParseError::Missing(node));
+            if first_only {
+                break;
+            }
+            did_visit_children = true;
+        }
+        if did_visit_children {
+            if cursor.goto_next_sibling() {
+                did_visit_children = false;
+            } else if cursor.goto_parent() {
+                did_visit_children = true;
+            } else {
+                break;
+            }
+        } else {
+            if cursor.goto_first_child() {
+                did_visit_children = false;
+            } else {
+                did_visit_children = true;
+            }
+        }
+    }
+    cursor.reset(tree.root_node());
+}
+
+// ------------------------------------------------------------------------------------------------
+// Types to package a tree and parse errors for that tree
+//
+// Parse errors contain `Node` values, that are parametrized by the lifetime `tree of the tree they
+// are part of. It is normally not possible to combine a value and references to that value in a single
+// data type. However, in the case of tree-sitter trees and nodes, the nodes do not point to memory in the
+// tree value, but both point to heap allocated memory. Therefore, moving the tree does not invalidate the
+// node. We use this fact to implement the TreeWithParseError* types.
+//
+// To be able to use these types in errors, we implement Send and Sync. These traits are implemented for
+// Tree, but not for Node. However, since the TreeWithParseError* types contain the tree as well as the nodes,
+// it is okay to implement Send and Sync.
+
+/// A type containing a tree and a parse error
+pub struct TreeWithParseError {
+    tree: Tree,
+    // the 'static lifetime is okay because we own `tree`
+    error: ParseError<'static>,
+}
+
+#[macro_export]
+macro_rules! new_tree_with_parse_error {
+    ($tree:expr,$error:expr) => {{
+        TreeWithParseError {
+            error: unsafe { std::mem::transmute($error) },
+            tree: $tree,
+        }
+    }};
+}
+
+impl TreeWithParseError {
+    pub fn tree(&self) -> &Tree {
+        &self.tree
+    }
+
+    pub fn into_tree(self) -> Tree {
+        self.tree
+    }
+
+    pub fn error(&self) -> &ParseError {
+        &self.error
+    }
+}
+
+impl std::fmt::Debug for TreeWithParseError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self.error)
+    }
+}
+
+// Send and Sync must be implemented for ParseError -> Node -> ffi::TSTree
+// This is okay because Send and Sync _are_ implemented for Tree, which also holds ffi::TSTree
+unsafe impl Send for TreeWithParseError {}
+unsafe impl Sync for TreeWithParseError {}
```

Comment:
It doesn't look like this type is used anywhere, can we drop it?  (If you'd rather keep it, remove the macro like you did with the types below)

Reply:
Forgot about that one, sorry about that. It is used in `TreeWithParseErrorOption::into_option`, which returns an `Option<TreeWithParseError`. It is a bit cumbersome to have three such similar types, but since users are stuck with what we provide them, I added this for completeness. I'll remove the macro.